src/parser/pattern.lisp

```diff
@@ -0,0 +1,120 @@
+(defpackage #:coalton-impl/parser/pattern
+  (:use
+   #:cl
+   #:coalton-impl/parser/base)
+  (:shadowing-import-from
+   #:coalton-impl/parser/base
+   #:parse-error)
+  (:local-nicknames
+   (#:cst #:concrete-syntax-tree)
+   (#:util #:coalton-impl/util))
+  (:export
+   #:pattern                            ; STRUCT
+   #:pattern-source                     ; ACCESSOR
+   #:pattern-list                       ; TYPE
+   #:pattern-var                        ; STRUCT
+   #:make-pattern-var                   ; ACCESSOR
+   #:pattern-var-name                   ; ACCESSOR
+   #:pattern-literal                    ; STRUCT
+   #:make-pattern-literal               ; CONSTRUCTOR
+   #:pattern-literal-value              ; ACCESSOR
+   #:pattern-wildcard                   ; STRUCT
+   #:make-pattern-wildcard              ; ACCESSOR
+   #:pattern-constructor                ; STRUCT
+   #:make-pattern-constructor           ; CONSTRUCTOR
+   #:pattern-constructor-name           ; ACCESSOR
+   #:pattern-constructor-patterns       ; ACCESSOR
+   #:parse-pattern                      ; FUNCTION
+   ))
+
+(in-package #:coalton-impl/parser/pattern)
+
+(defstruct (pattern
```

Comment:
after reading a good number of these files, and now onto this one, I have a general feeling that we have a lot of code that's clearly following some sort of "informal specification" that's not written down. If I didn't know Coalton codebase somewhat (and Coalton itself), I would have a hard time navigating everything.

It doesn't help that documentation is incredibly light. There is some helpful comments in long case analyses (e.g., about errors), but there's nothing as it pertains to what anything means, what invariants need to be satisfied, etc. etc.

I don't want to block this PR over it, but we need to spec something out somewhere, whether it's in the headers of these files, or somewhere else.